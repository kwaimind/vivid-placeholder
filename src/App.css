html {
  -webkit-box-sizing: border-box;
  -moz-box-sizing: border-box;
  box-sizing: border-box;
}
*,
*:before,
*:after {
  -webkit-box-sizing: inherit;
  -moz-box-sizing: inherit;
  box-sizing: inherit;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  text-rendering: optimizeLegibility;
}
body {
  background-color: #3d9970;
  font-family: Inter, -apple-system, BlinkMacSystemFont, Segoe UI, Helvetica,
    Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji, Segoe UI Symbol;
  font-weight: 400;
  color: #fff;
  line-height: 1.2;
  text-align: center;
  transition: background-color 0.5s ease;
}
.mono-text {
  font-family: SFMono-Regular, Consolas, Liberation Mono, Menlo, monospace;
}
h1,
span.emoji {
  font-size: 44px;
}
span.random-image {
  text-decoration: underline;
  cursor: pointer;
}
.intro {
  margin: 10px;
}
.app {
  min-height: 100vh;
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
}
.container {
  position: relative;
}
.button,
[type='file'] + label {
  color: #fff;
  background: #85144b;
  display: flex;
  overflow: hidden;
  margin: 10px;
  padding: 12px 12px;
  cursor: pointer;
  user-select: none;
  transition: all 150ms linear;
  text-align: center;
  white-space: nowrap;
  text-decoration: none !important;
  text-transform: none;
  text-transform: capitalize;
  border: 0 none;
  border-radius: 4px;
  font-size: 13px;
  font-weight: 500;
  line-height: 1.3;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  justify-content: center;
  align-items: center;
  flex: 0 0 160px;
  box-shadow: 0px 8px 8px rgba(0, 0, 0, 0.3);
}
.left {
  position: absolute;
  left: 0;
  bottom: 0;
  width: 130px;
  z-index: 9;
}
.right {
  position: absolute;
  right: 0;
  bottom: 0;
  width: 130px;
  background-color: #111111;
}
.button:hover {
  transition: all 150ms linear;
  opacity: 0.85;
}
button:active {
  transition: all 150ms linear;
  opacity: 0.75;
}
button:focus {
  outline: 1px dotted #959595;
  outline-offset: -4px;
}
.img-preview img {
  height: 90vh;
  box-shadow: 0px 8px 8px rgba(0, 0, 0, 0.3);
}
.swatch-overlay {
  position: absolute;
  bottom: 10px;
  left: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
}
.swatch-item {
  width: 100px;
  height: 100px;
  display: flex;
  justify-content: center;
  align-items: center;
  color: #111111;
}
.swatches {
  display: flex;
  justify-content: left;
  box-shadow: 0px 8px 8px rgba(0, 0, 0, 0.3);
}

[type='file'] {
  height: 0;
  overflow: hidden;
  width: 0;
}
[type='file'] + label:hover {
  background-color: darken(#f15d22, 10%);
  transform: scale(1.05);
}

@media screen and (max-width: 420px) {
  h1,
  span.emoji {
    font-size: 38px;
  }
  .intro {
    margin: 20px;
  }
  .img-preview img {
    width: 90vw;
    height: auto;
  }
  .swatch-overlay {
    bottom: 75px;
  }
  .swatch-item {
    width: calc(90vw / 6);
    height: calc(90vh / 12);
    font-size: 10px;
  }
}
