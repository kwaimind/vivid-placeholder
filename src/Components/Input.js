import React from 'react';
import PropTypes from 'prop-types';

export default function Input({ useUploadedImage }) {
  return (
    <>
      <input
        className="file-input"
        type="file"
        id="file"
        onChange={useUploadedImage}
      />
      <label htmlFor="file">choose a file</label>
    </>
  );
}

Input.propTypes = {
  useUploadedImage: PropTypes.func.isRequired,
};
