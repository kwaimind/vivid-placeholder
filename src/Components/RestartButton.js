import React from 'react';
import PropTypes from 'prop-types';

function RestartButton({ handleRestart }) {
  return (
    <div
      className="button right"
      onClick={handleRestart}
      onKeyDown={handleRestart}
      role="form"
    >
      Try a new image
    </div>
  );
}

function shouldUpdate(prevProps, nextProps) {
  return prevProps !== nextProps;
}

RestartButton.propTypes = {
  handleRestart: PropTypes.func.isRequired,
};

export default React.memo(RestartButton, shouldUpdate);
