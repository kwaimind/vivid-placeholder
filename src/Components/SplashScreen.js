import React from 'react';
import PropTypes from 'prop-types';

export default function SplashScreen({ useRandomImage }) {
  return (
    <>
      <span className="emoji" role="img" aria-label="camera">
        ðŸ“·
      </span>
      <h1>Vivid Placeholder</h1>
      <p>
        Upload an image to find it&apos;s dominant color and swatches.
        <br />
        Or take a &nbsp;
        <span className="random-image" role="button" onClick={useRandomImage}>
          random Unsplash image.
        </span>
      </p>
    </>
  );
}

SplashScreen.propTypes = {
  useRandomImage: PropTypes.func.isRequired,
};
