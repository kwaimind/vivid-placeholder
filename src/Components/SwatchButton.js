/* eslint-disable react/destructuring-assignment */
import React from 'react';
import PropTypes from 'prop-types';

export default function SwatchButton(props) {
  const { showSwatches, setShowSwatches } = props.actions;
  return (
    <div className="button left" onClick={() => setShowSwatches(!showSwatches)}>
      {showSwatches ? 'Hide' : 'Show'}
      Swatches
    </div>
  );
}

SwatchButton.propTypes = {
  //
};
