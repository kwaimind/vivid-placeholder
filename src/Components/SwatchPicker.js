import React, { useContext } from 'react';

import AppContext from '../Context/AppContext';

export default function SwatchPicker() {
  return (
    <div className="swatch-overlay">
      <div className="swatches">
        <RenderSwatches />
      </div>
    </div>
  );
}

function RenderSwatches() {
  const colors = useContext(AppContext);
  return colors.slice(0, 6).map((color) => (
    <div
      key={color.toString()}
      className="swatch-item mono-text"
      style={{
        backgroundColor: color,
      }}
    >
      {color}
    </div>
  ));
}
