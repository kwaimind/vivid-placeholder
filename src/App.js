import React, { useState, useEffect } from 'react';
import { ColorExtractor } from 'react-color-extractor';
import axios from 'axios';

import { introColors } from './Constants';
import SwatchButton from './Components/SwatchButton';
import RestartButton from './Components/RestartButton';
import SwatchPicker from './Components/SwatchPicker';
import SplashScreen from './Components/SplashScreen';

import './App.css';

export default function App() {
  const [colors, setcolors] = useState([]);
  const [restart, setRestart] = useState(false);
  const [imageUrl, setimageUrl] = useState('');
  const [fileName, setfileName] = useState('');
  const [fileType, setfileType] = useState('');
  const [swatchOverlay, setswatchOverlay] = useState(false);

  useEffect(() => {
    document.body.style.backgroundColor =
      introColors[Math.floor(Math.random() * introColors.length)];
  }, [restart]);

  const handleRestart = () => {
    setRestart(!restart);
    setimageUrl('');
    setfileName('');
    setfileType('');
    setcolors([]);
    setRestart(!restart);
  };

  const useRandomImage = async () => {
    try {
      const getImage = await axios.get('https://source.unsplash.com/random');
      setimageUrl(getImage.request.responseURL);
    } catch (error) {
      console.error(error);
    }
  };

  const useUploadedImage = (e) => {
    e.preventDefault();
    let reader = new FileReader();
    let file = e.target.files[0];
    reader.onloadend = () => {
      setimageUrl(URL.createObjectURL(file));
      setfileType(file.type);
      setfileName(file.name);
    };
    reader.readAsDataURL(file);
  };

  const extractColors = () => {
    if (imageUrl) {
      return <ColorExtractor src={imageUrl} getColors={handleColors} />;
    }
  };

  const handleColors = (colors) => {
    setcolors([...colors, ...colors]);
    document.body.style.backgroundColor = colors[0];
  };

  const renderSwatches = () => {
    return colors.slice(0, 6).map((color, id) => {
      return (
        <div
          key={id}
          className="swatch-item mono-text"
          style={{
            backgroundColor: color,
          }}
        >
          {color}
        </div>
      );
    });
  };

  const swatchOverlaySwitch = () => {
    swatchOverlay ? setswatchOverlay(false) : setswatchOverlay(true);
  };

  return (
    <div className="app">
      {!imageUrl && (
        <div className="intro">
          <SplashScreen useRandomImage={useRandomImage} />

          <input
            className="file-input"
            type="file"
            id="file"
            onChange={useUploadedImage}
          />
          <label htmlFor="file">choose a file</label>
        </div>
      )}

      {extractColors()}

      {colors.length > 0 ? (
        fileType.includes('image') ? (
          <div>
            <div className="img-preview">
              <img src={imageUrl} alt={fileName} />
            </div>
            {swatchOverlay && <SwatchPicker renderSwatches={renderSwatches} />}
            <SwatchButton toggle={swatchOverlay} action={swatchOverlaySwitch} />
            <RestartButton handleRestart={handleRestart} />
          </div>
        ) : (
          <p>Whoops. You need to upload an image for this to work.</p>
        )
      ) : null}
    </div>
  );
}
