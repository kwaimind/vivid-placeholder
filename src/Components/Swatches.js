import React, { useState } from 'react';

import SwatchButton from './SwatchButton';
import SwatchPicker from './SwatchPicker';

export default function Swatches({ colors }) {
  const [showSwatches, setShowSwatches] = useState(false);

  return (
    <>
      <SwatchButton toggle={showSwatches} action={setShowSwatches} />
      {showSwatches && <SwatchPicker colors={colors} />}
    </>
  );
}
