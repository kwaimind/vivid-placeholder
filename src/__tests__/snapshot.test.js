import React from 'react';
import renderer from 'react-test-renderer';
import App from '../App';
import Input from '../Components/Input';
import RestartButton from '../Components/RestartButton';
import SplashScreen from '../Components/SplashScreen';
import SwatchButton from '../Components/SwatchButton';
import Swatches from '../Components/Swatches';
import SwatchPicker from '../Components/SwatchPicker';

const fakeFunc = () => {
  return;
};

describe('The components match the previous snapshot', () => {
  it('<App />', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('<Input />', () => {
    const tree = renderer
      .create(<Input useUploadedImage={fakeFunc} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('<RestartButton />', () => {
    const tree = renderer
      .create(<RestartButton handleRestart={fakeFunc} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('<SplashScreen />', () => {
    const tree = renderer
      .create(<SplashScreen useRandomImage={fakeFunc} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('<Swatches />', () => {
    const tree = renderer.create(<Swatches />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('<SwatchButton />', () => {
    const tree = renderer
      .create(
        <SwatchButton
          actions={{ showSwatches: fakeFunc, setShowSwatches: fakeFunc }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('<SwatchPicker />', () => {
    const tree = renderer.create(<SwatchPicker />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
