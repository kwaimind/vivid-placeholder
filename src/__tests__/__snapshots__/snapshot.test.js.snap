// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The components mount with no errors <App /> renders without crashing 1`] = `
<div
  className="app"
>
  <div
    className="intro"
  >
    <span
      aria-label="camera"
      className="emoji"
      role="img"
    >
      ðŸ“·
    </span>
    <h1>
      Vivid Placeholder
    </h1>
    <p>
      Upload an image to find it's dominant color and swatches.
      <br />
      Or take a Â 
      <span
        className="random-image"
        onClick={[Function]}
        role="button"
      >
        random Unsplash image.
      </span>
    </p>
    <input
      className="file-input"
      id="file"
      onChange={[Function]}
      type="file"
    />
    <label
      htmlFor="file"
    >
      choose a file
    </label>
  </div>
  
</div>
`;

exports[`The components mount with no errors <Input /> renders without crashing 1`] = `
Array [
  <input
    className="file-input"
    id="file"
    onChange={[Function]}
    type="file"
  />,
  <label
    htmlFor="file"
  >
    choose a file
  </label>,
]
`;

exports[`The components mount with no errors <RestartButton /> renders without crashing 1`] = `
<div
  className="button right"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="form"
>
  Try a new image
</div>
`;

exports[`The components mount with no errors <SplashScreen /> renders without crashing 1`] = `
Array [
  <span
    aria-label="camera"
    className="emoji"
    role="img"
  >
    ðŸ“·
  </span>,
  <h1>
    Vivid Placeholder
  </h1>,
  <p>
    Upload an image to find it's dominant color and swatches.
    <br />
    Or take a Â 
    <span
      className="random-image"
      onClick={[Function]}
      role="button"
    >
      random Unsplash image.
    </span>
  </p>,
]
`;

exports[`The components mount with no errors <SwatchButton /> renders without crashing 1`] = `
<div
  className="button left"
  onClick={[Function]}
>
  Hide
  Swatches
</div>
`;

exports[`The components mount with no errors <SwatchPicker /> renders without crashing 1`] = `
<div
  className="swatch-overlay"
>
  <div
    className="swatches"
  />
</div>
`;

exports[`The components mount with no errors <Swatches /> renders without crashing 1`] = `
<div
  className="button left"
  onClick={[Function]}
>
  Show
  Swatches
</div>
`;
